### PR TITLE
Add more nil pointer check for CSI related code in backup controller

### PR DESCRIPTION
**Pull request by blackpiglet**
Add some corner cases checking for CSI snapshot in backup controller.

Signed-off-by: Xun Jiang <blackpiglet@gmail.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes https://github.com/vmware-tanzu/velero/issues/5207
Fixes https://github.com/vmware-tanzu/velero/issues/5244

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/controller/backup_controller.go [hunk 1]
```diff
@@ -969,6 +979,9 @@ func (c *backupController) deleteVolumeSnapshot(volumeSnapshots []snapshotv1api.
 				if vsc.Spec.DeletionPolicy == snapshotv1api.VolumeSnapshotContentDelete {
 					modifyVSCFlag = true
 				}
+			} else {
+				logger.Errorf("VS %s/%s is not ready. This is not expected.", vs.Namespace, vs.Name)
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:983]: VS-> VolumeSnapshot
change log msg in line 1010 from "Deleting VolumeSnapshotContent %s " to "Deleting VolumeSnapshot ..."

File: pkg/controller/backup_controller.go [hunk 2]
```diff
@@ -969,6 +979,9 @@ func (c *backupController) deleteVolumeSnapshot(volumeSnapshots []snapshotv1api.
...
+				return
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:984]: Don't return, we should still try to delete the VS.

File: pkg/controller/backup_controller.go [hunk 3]
```diff
@@ -670,6 +670,16 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 		if err != nil {
 			backupLog.Errorf("fail to wait VolumeSnapshot change to Ready: %s", err.Error())
 		}
+
+		// checkVolumeSnapshotReadyToUse could run for a long time. Update VolumeSnapshot array after it finishes.
+		err = c.kbClient.List(context.Background(), vsList, &kbclient.ListOptions{LabelSelector: selector})
+		if err != nil {
+			backupLog.Error(err)
+		}
+		if len(vsList.Items) >= 0 {
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:679]: No need to check the len(x) >= 0 and we should put it in the `else` clause
Same issue in line 666.

File: pkg/controller/backup_controller.go [hunk 4]
```diff
@@ -670,6 +670,16 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 		if err != nil {
 			backupLog.Errorf("fail to wait VolumeSnapshot change to Ready: %s", err.Error())
 		}
+
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:673]: We should move the code to list volumesnapshots and assigning the list to backup request  into `c.checkVolumeSnapshotReadyToUse`

 btw rename it to `waitVolumeSnapshotReadyToUse` would be better.

**Comment by reasonerjt** [comment]
@blackpiglet Please combine all the nil-pointer checking in this PR and make sure this is cherry-picked to v1.9.3

**Comment by blackpiglet** [comment]
Tried to add a UT case for function `waitVolumeSnapshotReadyToUse`, but didn't make the VolumeSnapshot informer Lister work in UT, so modified according to Yonghui's suggestion. Will add the UT case later.

**Review by Lyndon-Li — DISMISSED** [review]

**Review by qiuming-best — APPROVED** [review]

File: pkg/controller/backup_controller.go [hunk 5]
```diff
@@ -942,7 +963,16 @@ func (c *backupController) checkVolumeSnapshotReadyToUse(ctx context.Context, vo
 			return err
 		})
 	}
-	return eg.Wait()
+
+	err := eg.Wait()
```
Comment (ywk253100) [inline pkg/controller/backup_controller.go:967]: Check error?

Comment (blackpiglet) [reply pkg/controller/backup_controller.go:967]: Error is needed for return and record. No need for handling.

**Review by ywk253100 — APPROVED** [review]

**Pull request merged**